esmvalcore/_recipe/to_datasets.py

```diff
@@ -0,0 +1,439 @@
+    """
+    _check_ancillaries_valid(var_facets)
+    _check_ancillaries_valid(ds_facets)
+    merged = {}
+    for facets in var_facets:
+        merged[facets['short_name']] = facets
+    for facets in ds_facets:
+        short_name = facets['short_name']
+        if short_name not in merged:
+            merged[short_name] = {}
+        merged[short_name].update(facets)
+    return list(merged.values())
+
+
+def _get_facets_from_recipe(
+    variable_group: str,
+    recipe_variable: dict[str, Any],
+    recipe_dataset: dict[str, Any],
+    session: Session,
+) -> tuple[dict[str, Any], list[dict[str, Any]]]:
+    """Read the facets for a single dataset definition from the recipe."""
+    facets = deepcopy(recipe_variable)
+    recipe_dataset = deepcopy(recipe_dataset)
+
+    ancillaries = _merge_ancillary_dicts(
+        facets.pop('ancillary_variables', []),
+        recipe_dataset.pop('ancillary_variables', []),
+    )
+
+    facets.update(recipe_dataset)
+
+    if 'short_name' not in facets:
+        facets['short_name'] = variable_group
+
+    # Flaky support for limiting the number of years in a recipe.
+    # If we want this to work, it should actually be done based on `timerange`,
+    # after any wildcards have been resolved.
+    if 'end_year' in facets and session['max_years']:
+        facets['end_year'] = min(
+            facets['end_year'],
+            facets['start_year'] + session['max_years'] - 1)
+
+    # Legacy: support start_year and end_year instead of timerange
+    _replace_years_with_timerange(facets)
+
+    # Legacy: support wrong capitalization of obs4MIPs
+    if facets['project'] == 'obs4mips':
+        logger.warning("Correcting capitalization, project 'obs4mips' "
+                       "should be written as 'obs4MIPs'")
+        facets['project'] = 'obs4MIPs'
+
+    check.variable(
+        facets,
+        required_keys=(
+            'short_name',
+            'mip',
+            'dataset',
+            'project',
+        ),
+    )
+
+    return facets, ancillaries
+
+
+def _get_all_datasets_for_variable(
+    diagnostic: str,
+    variable_group: str,
+    recipe_variable: dict[str, Any],
+    recipe_datasets: Iterable[dict[str, Any]],
+    session: Session,
+) -> list[Dataset]:
+    """Read the datasets from a variable definition in the recipe."""
+    datasets = []
+    idx = 0
+
+    for recipe_dataset in recipe_datasets:
+        facets, ancillaries = _get_facets_from_recipe(
+            variable_group=variable_group,
+            recipe_variable=recipe_variable,
+            recipe_dataset=recipe_dataset,
+            session=session,
+        )
+
+        dataset = Dataset(**facets)
+        dataset.session = session
+        dataset['diagnostic'] = diagnostic
+        dataset['variable_group'] = variable_group
+
+        for dataset1 in dataset.from_ranges():
+            for ancillary_facets in ancillaries:
+                dataset1.add_ancillary(**ancillary_facets)
```

Comment:
If the experiment in the variable is given as a list (`exp: [historical, ssp245`), this also automatically updates the experiment for the ancillary variables. If their mip is in `fx`, the tool ends up crashing when trying to concatenate ancillaries from different experiments because there is no time coordinate.

Reply:
Is that when you're using `exp: '*'` for the ancillary?

Reply:
No, it only happens if you define the ancillary with just the mip and the short name:

```
    ancillary_variables:
      - short_name: areacella
        mip: fx
```

Reply:
Yes, I can see why that doesn't work. If you want a specific experiment, you need to specify it.

```yaml
    ancillary_variables:
      - short_name: areacella
        mip: fx
        exp: historical
```

Reply:
Ok, maybe it's something that could be fixed at some point though. To avoid concatenation of fx files. But it maybe for another PR, this one is big enough.

Reply:
Isn't that the same issue I encountered in [1877](https://github.com/ESMValGroup/ESMValCore/pull/1877#issuecomment-1377249198) (see number 5 in this list)? From the discussion there I understood that this will be fixed in the recipe, but apparently it isn't. I really think we should fix that here.

Especially if you want to use the automatic addition of cell measures (e.g., by the usage of `area_statistics` with no explicit ancillaries specified, see @sloosvel's comment above) it would be really annoying if you have to specify an ancillary variable for every dataset just because you want to use multiple experiments.

Reply:
Fixed in [301903e](https://github.com/ESMValGroup/ESMValCore/pull/1609/commits/301903eb9213501a9f392b85bd7198763b1c5ae3)